code/object/objectdock.cpp

```diff
@@ -337,6 +339,107 @@ float dock_calc_docked_speed(object *objp)
 	return dfi.maintained_variables.float_value;
 }
 
+void dock_calc_total_moi(matrix* dest, object* objp, vec3d *center)
+{
+	Assertion(dest != nullptr, "dock_calc_total_moi invalid dest");
+	Assertion(objp != nullptr, "dock_calc_total_moi invalid objp");
+
+	*dest = vmd_zero_matrix;
+
+	dock_function_info dfi;
+	dfi.parameter_variables.vecp_value = center;
+	dfi.maintained_variables.matrix_value = dest;
+
+	dock_evaluate_all_docked_objects(objp, &dfi, dock_calc_total_moi_helper);
+}
+
+// This ship is the only ship NOT moved by docking AI to keep everyone together
+// All the other ships in the tree will update based on this one
+// Since this is based on current speed don't expect it to remain consistent between frames
+object* dock_find_dock_root(object *objp)
+{
+	Assertion(objp != nullptr, "dock_find_dock_root invalid argument");
+
+	dock_function_info dfi;
+	object* fastest_objp;
+
+	dfi.maintained_variables.objp_value = nullptr;
+
+	// find the object with the highest speed
+	dock_evaluate_all_docked_objects(objp, &dfi, dock_find_max_speed_helper);
+	fastest_objp = dfi.maintained_variables.objp_value;
+
+	// if we have no max speed, just use the given one
+	if (fastest_objp == nullptr)
+		fastest_objp = objp;
+
+	return fastest_objp;
+}
+
+void dock_calculate_whack_docked_object(vec3d* force, const vec3d* rel_world_hit_pos, object* objp)
```

Comment:
Rename this to `dock_calculate_and_apply_whack_docked_object`, per comment on `physics_calculate_whack`